### PR TITLE
NIFI-13289 Add tooltip to NewCanvas item

### PR DESCRIPTION
**Pull request by zeevo**
# Summary

[NIFI-13289](https://issues.apache.org/jira/browse/NIFI-13289)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 21

### UI Contributions

- [x] NiFi is modernizing its UI. Any contributions that update the [current UI](https://github.com/apache/nifi/tree/main/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui) also need to be implemented in the [new UI](https://github.com/apache/nifi/tree/main/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi).  

### Licensing

- [x] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [x] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [x] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/flow-designer/ui/canvas/header/new-canvas-item/new-canvas-item.component.ts [hunk 1]
```diff
@@ -25,19 +25,33 @@ import { selectDragging } from '../../../../state/flow/flow.selectors';
...
     @Input() type!: ComponentType;
     @Input() iconClass = '';
     @Input() iconHoverClass = '';
     @Input() disabled = false;
+    @Input() tooltip = '';
+
+    protected readonly TextTip = TextTip;
+
+    tooltipPosition: ConnectedPosition = {
+        originX: 'center', // maybe start?
```
Comment (zeevo) [inline nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/flow-designer/ui/canvas/header/new-canvas-item/new-canvas-item.component.ts:49]: I need to remove this comment

**Review by rfellows — CHANGES_REQUESTED** [review]
@zeevo, Thanks for the contribution.

There is a minor issue with this implementation that I think we can resolve. The main problem is that the tooltip doesn't go away while dragging and tooltips are shown as you drag over other component types in the header.

Kapture 2024-05-28 at 13 05 13


One possible option could be to implement a new `@Input()` for `nifiTooltipDisbled` (borrowing from how angular material does it with `matTooltipDisabled`). When disabled, the overlay wouldn't get rendered.
Then, in the template that uses the nifiTooltip set it to be true while dragging. There is already a member variable for this.

You will most likely need to short circuit the `mouseEnter` method if disabled so it doesn't attach the overlay. Probably also need to consider if the overlay has already been attached and the mouse is still over the element. It should get detached in this case. You can probably track this in a new mouseMove method and consider the attached state of the overlay and if it is disabled or not.

**Comment by zeevo** [comment]
@rfellows Thanks for the review and in-depth suggestions. It really helped.

My latest commit has the following new behaviors:
- Tooltip dismisses on drag.
- Tooltip reappears on mouse up if within a canvas item.

I decided to short circuit within `attach()` as it made things easier in my opinion. Let me know if you disagree with that implementation, and I'll try doing it within the `HostListener` event methods.

File: nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/ui/common/tooltips/nifi-tooltip.directive.ts [hunk 2]
```diff
@@ -58,11 +59,29 @@ export class NifiTooltipDirective<T> implements OnDestroy {
         }
     }
 
+    @HostListener('mousemove')
+    mouseMove() {
+        if (this.overlayRef?.hasAttached() && this.tooltipDisabled) {
+            this.overlayRef?.detach();
+        } else {
+            this.attach();
```
Comment (rfellows) [inline nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/ui/common/tooltips/nifi-tooltip.directive.ts:67]: I don't think this attach is needed. In fact it causes console errors...

Screenshot 2024-05-29 at 8 49 10 AM

File: nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/ui/common/tooltips/nifi-tooltip.directive.ts [hunk 3]
```diff
@@ -58,11 +59,29 @@ export class NifiTooltipDirective<T> implements OnDestroy {
...
+        }
+    }
+
+    @HostListener('mouseup')
+    mouseup() {
+        this.attach();
```
Comment (rfellows) [inline nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/ui/common/tooltips/nifi-tooltip.directive.ts:73]: Nice touch but i think this should also be guarded. It currently can cause an error (similar to the above) if you mousedown outside of the target object and then mouseup over it. The tooltip will render based on the mouseenter and then try to attach another on the mouseup.
```suggestion
        if (!this.overlayRef?.hasAttached()) {
            this.attach();
        }
```

**Review by rfellows — CHANGES_REQUESTED** [review]
@zeevo This is looking pretty good. Just a couple of things to clean up and then I think we can get it merged.

File: nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/ui/common/tooltips/nifi-tooltip.directive.ts (hunk 3 above)
Comment (zeevo) [reply nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/ui/common/tooltips/nifi-tooltip.directive.ts]: Added guard

**Review by rfellows — APPROVED** [review]
This looks good 👍 . Thanks for the contribution, @zeevo.
I'll get this merged once CI completes.

**Pull request merged**